packages/realm/src/Object.ts

```diff
@@ -236,25 +228,39 @@ export class RealmObject<T = DefaultObject> {
   }
 
   /**
+   * @internal
    * The Realm managing the object.
+   * Note: this is on the injected prototype from ClassMap.defineProperties().
    */
-  public readonly realm!: Realm;
+  public declare readonly [REALM]: Realm;
```

Comment:
When new people are joining the team, we are likely required to teach them a bit of TypeScript :smile:

I am not sure how to capture it. Should we have a file under `contrib` with tips-n-tricks, policies, and best practises? Or is a comment in the code a better way?

Reply:
Perhaps it could be good to have it in both places (contrib and comments). I think having it in comments where it's used would probably be a better reminder for developers.

Reply:
The inline comment is more important for the collections case (which I haven't put up for PR yet, and may not until I get back in Jan), because in that case we actually are calling the constructor. For Object, the only time the constructor is ever invoked is in the `super()` call of a subclass when `new UserType()` is used for insertion, and even in that case we immediately throw the object away. Plus I think we plan to get rid of that syntax as a way of making objects.

This (almost) doesn't feel like a typescript thing to me, since the "weird" behavior is in the JS layer. It sort-of is because TS had the class field declaration syntax for a long time before it was added to JS. And they change their default behavior when [targeting ES2022 or higher](https://www.typescriptlang.org/tsconfig#useDefineForClassField). However, this follows the normal rules for the `declare` keyword which means that something is only intended to inform the typesystem of a fact rather than to emit any JS code. The weird thing is that for so long class fields never emitted JS code, but under newer targets, it does seem reasonable for them to use class fields as something that gets emitted into JS. Especially since it can provide a perf boost in some cases (with newer JS engines [where it doesn't trigger deoptimizations](https://v8.dev/blog/faster-class-features)).

I think we should have a policy of something vagly like "Use JS-level class fields for 'normal' fields that are added using normal `this.field = blah` syntax to 'normal' (aka non-proxy) objects. Use `declare` for class members that are added using 'magic' such as being added to a prototype chain, or set on `this` in the `super()` call, or whenever `super()` returns a Proxy." I have no idea where we should put this bit of tribal knowledge though. Do we have any sort of style guide or local best practices document? It feels like that is where this kind of stuff lives in other projects I've seen.